add brief description of common extension points (#53)
colcon/colcon.readthedocs.org: developer/extension-point.rst

Extension points
================

> The following describes some of the extension points and how they are being used.

Reviewer: ```suggestion
The following describes some of the extension points and how they are used.
```

> VerbExtensionPoint

Reviewer 2: Worthwhile to link these to the actual code in question, eg with `:class:`?

Author: Unfortunately we don't have generated API docs at the moment. Once that is the case it would be great to link to it.

Reviewer 2: Oh, ha. That's right, I'm forgetting how Sphinx actually works...

Each verb extension can add command line arguments to the verb specific subparser in ``argparse``.

PackageDiscoveryExtensionPoint
------------------------------

This extension point is used by verb extensions.
Each package discovery extension can add command line arguments which are grouped in the "Discovery arguments" section of the usage help.
Each extension returns a set of package descriptors which identify a package with the tuple path, name, and type.
While a package discovery extension only provides candidate paths it uses the package identification extension point to determine if a path contains a package.

If explicit arguments are provided for any package discovery extensions only those extension are being used.
Otherwise the package discovery extensions which provide a default value are used.

PackageIdentificationExtensionPoint
-----------------------------------

This extension point is used by package discovery extensions.
A package identification extension determines if a given path contains a package and returns the tuple path, name, and type.

Package identification extensions are grouped by their priority.

> The extensions are being invoked in order of their priority and the first extension successfully identifying a package stops further extensions from being considered.

Reviewer: ```suggestion
The extensions are invoked in order of their priority and the first extension successfully identifying a package stops further extensions from being considered.
```

> If multiple extensions have the same priority they are all being invoked.

Reviewer: ```suggestion
If multiple extensions have the same priority they are all invoked.
```

Each task extension implements the logic for the combination of a ``verb`` and a ``package type``, e.g. building a Python package.

In order to be invoked a task extension needs parameters provided by a context object.
A job object brings those two together.

ExecutorExtensionPoint
----------------------

This extension point is used by some verb extensions to execute a set of jobs.
The executor extension with the highest priority is used by default.

ShellExtensionPoint
-------------------

This extension point is used by some task extensions to generate shells scripts to setup the environment or create environment hooks for each supported shell.
Each extension implements the logic for a specific shell.
For the primary shells each extension also provides the environment to run commands.

EnvironmentExtensionPoint
-------------------------

This extension point is used by some task extensions as well as shell extensions to create environment hooks for each supported shell.
Each extension creates environment hooks for a specific environment variable and uses the shell extensions to generate scripts for each supported shell.

EventHandlerExtensionPoint
--------------------------

> This extension point defines how events are being handled.

Reviewer: ```suggestion
This extension point defines how events are handled.
```

> An event can be any kind of object, like a line of output, a job being started, progressing or ended, a command being executed.

Reviewer: ```suggestion
An event can be any kind of object, like a line of output, a job starting, progressing or ending, or a command executing.
```

> Every event is being dispatched to all extensions in order of their priority.

Reviewer: ```suggestion
Every event is dispatched to all extensions in order of their priority.
```
